lxd/device/device.go

```diff
@@ -0,0 +1,74 @@
+package device
+
+import (
+	"fmt"
+)
+
+// ErrUnsupportedDevType is the error that occurs when an unsupported device type is created.
+var ErrUnsupportedDevType = fmt.Errorf("Unsupported device type")
+
+// Config represents user defined persistent config for a device.
+type Config map[string]string
+
+// VolatileConfig represents LXD defined persistent config for a device.
+type VolatileConfig map[string]string
+
+// RunConfig represents LXD defined run-time config used for device setup.
+type RunConfig map[string]string
+
+// Device represents a device that can be added to an instance.
+type Device interface {
+	// Start peforms any host-side configuration required to setup the device for the instance.
+	// This can be when a device is plugged into a running instance or the instance is starting.
+	// VolatileConfig from previous instantiations of this device is required.
+	// It returns run-time configuration necessary for configuring the instance with the new
+	// device and any changes to VolatileConfig that need to be persisted.
+	Start(VolatileConfig) (RunConfig, VolatileConfig, error)
+
+	// Stop performs any host-side cleanup required when a device is removed from an instance,
+	// either due to unplugging it from a running instance or instance is being shutdown.
+	// VolatileConfig from previous instantiations of this device is required.
+	// It return sany changes to VolatileConfig that need to be persisted.
+	Stop(VolatileConfig) (VolatileConfig, error)
+
+	// setConfig stores the contents of instanceType and Config into device.
+	setConfig(string, Config)
+
+	// validate is used by New() to check for per-device type validation.
+	validate() error
```

Comment:
Pretty sure that one is meant to be exported

Reply:
@stgraber happy to make that exported, although my thinking around that was to make the `device.New()` function call `validate()` as being the only way to validate a function rather than having to perform initialisation on the struct first (such as setting instance type and config, as I was thinking that validation may well depend on both the instance type and the device config). I am also going to make `device.New()` return a pointer to the Device so that it will be impossible to call any of the device's exported functions without New() parsing validation and not returning an error.

Reply:
@stgraber moved the validate and init (renamed from setConfig) to an internal only interface, so that the main Device interface only contains exported functions. Updated comments to explain this too.